---
title: "Microbiome Browser"
output: 
  flexdashboard::flex_dashboard:
    vertical_layout: fill
    social: "menu"
    theme:
       version: 4
       bootswatch: pulse
runtime: shiny_rmd
---

```{r}
library(flexdashboard)
library(knitr)
library(rpivotTable)
library(openintro)
library(highcharter)
library(ggvis)
library(heatmaply)
library(thematic)
library(bslib)

#thematic_on(bg= "black", fg="white", accent = "red", font = "Indie Flower")

func_env <- new.env()
source(here::here('microbiome_functions.R'), local = func_env)
attach(func_env)
```

```{r set_variables}
# make columns with a combination of no more than 3 factors
# And run ANCOM-BC or 

analysis_name <- "Soil_taxonomy"
sample_type <- "Soil"

otu_table <- here::here("16S_otu_table.biom")
plot_dir <- here::here()
stringToReplace <- c("unclassified", "Other", "NA")
suffix <- c(";_",";_", ";_")
metadata_file <- here::here('metadata.tsv')


# Groups used for alpha diversity statistics, box plots and bar plots
groups <- c("Treatment")
names(groups) <- groups
x_labs <- c("Treatment")

names(x_labs) <- x_labs


# Groups used for testing interactions and for betaDiversity ploting
main_groups <- c("Treatment") 

# Setting the colors to represent the levels of all the factor variables
Treatment_colors <- c(CS="aquamarine", SA= "blue", SL= "gold",
                        SLSA="purple", CMP="saddlebrown",
                        CMR="black", GP="ivory3")

annotation_colors <- list(Treatment = Treatment_colors)

# Factor variables
# Main effects
Treatment_order <- c("CS", "SA", "SL", "SLSA", "CMP", "CMR", "GP")
samples2drop <- c("S43")
```

```{r prepare_metadata, include=TRUE}
# Prepare the metadata

COLNAMES <- c("Sample_name", "BarcodeSequence", "LinkerPrimerSequence",
              "InputFileName", "TreatmentName", "Replicate", "Application",
              "SampleType", "Description" )

columns2keep <- c("Sample_name", groups)

metadata <-  read_delim(file = metadata_file, delim = "\t",
                        col_names = COLNAMES, skip = 1) %>%
  as.data.frame(stringAsFactor=FALSE) %>% 
  mutate(Application= case_when(
    Application == "Compost" ~ "CMP",
    Application == "Coumarin" ~ "CMR",
    Application == "Gypsum" ~ "GP",
    Application == "Control" ~ "CS",
    TRUE ~ Application
  )) %>% 
  rename(Treatment=Application) %>% 
  filter(!Sample_name %in% samples2drop) %>% 
  filter(str_detect(Treatment, pattern="Expt", negate=TRUE)) %>% 
  filter(SampleType == sample_type) %>% 
  mutate(Treatment=factor(Treatment, levels=Treatment_order)) %>%
  arrange(match(x = Treatment,table =levels(Treatment)))


rownames(metadata) <- metadata$Sample_name


for(column in colnames(metadata)){
  if(is.character(metadata[,column])){
    metadata[,column]<- tidyr::replace_na(metadata[,column], 'NONE')
    metadata[,column] <- gsub(pattern = "NA ", replacement = '',
                              x = metadata[,column])
  }
}

sample_order <- metadata %>% pull(Sample_name)
factors_order <-  list("Sample_name" = sample_order,
                       "Treatment" = Treatment_order)

# Set factor order
for(factor_column in names(factors_order)){
  
  metadata[,factor_column] <- factor(x = metadata[,factor_column], 
                                     levels = factors_order[[factor_column]])
  
}

original_metadata <- metadata
#create_dt(metadata)
metadata <- metadata[,columns2keep]
metadata <- metadata %>% 
  rownames_to_column(var = "Sample.ID") %>% 
  as.data.frame()

rownames(metadata) <- metadata[,"Sample.ID"]

```


```{r read_tables}
library(biomformat)
# read in the ASV table
feature_table <- read_biom(otu_table) %>% 
  biom_data %>%  as.matrix()
rownames(feature_table) <- paste0("denovo",rownames(feature_table))

colnames(feature_table) <- colnames(feature_table) %>%  
  str_replace_all("\\.16S", "")

# Filter out rare OTUS
#threshold <- sum(feature_table) * 0.00005

#feature_table <- feature_table[rowSums(feature_table) >= threshold,]

raw_taxonomy_table <- read_biom(otu_table) %>% observation_metadata() 

raw_taxonomy_table$Feature.ID <- paste0("denovo",rownames(raw_taxonomy_table))
raw_taxonomy_table$Confidence <- 0.9

raw_taxonomy_table  <- raw_taxonomy_table %>% dplyr::mutate(Taxon=str_c(
                                                 taxonomy1, taxonomy2, 
                                                 taxonomy3, taxonomy4, 
                                                 taxonomy5, taxonomy6,
                                                 taxonomy7, sep = ";")) %>% 
  dplyr::select(Feature.ID, Taxon, Confidence)

#raw_taxonomy_table <-  read_delim(file = taxonomy, delim = "\t")
rownames(raw_taxonomy_table) <- raw_taxonomy_table$Feature.ID

taxa2filter <- raw_taxonomy_table[grep(pattern = "Chloroplast|Mitochondria",
                        x = raw_taxonomy_table$Taxon, ignore.case = T),] %>%
  rownames()
taxonomy_table <- parse_taxonomy(raw_taxonomy_table)
taxonomy_table <- taxonomy_table[rownames(feature_table),]
asv_names <- rownames(taxonomy_table)
taxonomy_table <- process_metaphlan_taxonomy(taxonomy_table)
taxonomy_table <- fix_names(taxonomy=taxonomy_table,
                            stringToReplace=stringToReplace,suffix=suffix)
rownames(taxonomy_table) <- asv_names

# Drop Chloroplast and Mitochondrial sequences
feature_table <- feature_table[!(rownames(feature_table) %in% taxa2filter),]

taxonomy_table <- taxonomy_table[!(rownames(taxonomy_table) %in% taxa2filter),]
# Prepare feature tables
taxon_levels <- colnames(taxonomy_table)
names(taxon_levels) <- taxon_levels
taxon_tables <- map(.x = taxon_levels,
                    .f = make_feature_table,
                    count_matrix = feature_table, 
                    taxonomy = taxonomy_table, 
                    samples2keep=sample_order)

```


General Information
=====================================

Row
-------------------------------------

### Metadata

```{r}
create_dt(original_metadata)
```

Row
-------------------------------------

### Number of samples per category

```{r}
categories_freq_df <- metadata %>% 
  mutate(Sample_name=as.character(Sample_name)) %>% 
  summary(maxsum=100) %>% 
  as.data.frame() 

samples_per_category_df <- categories_freq_df[complete.cases(categories_freq_df),] %>% 
  dplyr::select(Var2:Freq) %>% 
  dplyr::filter(str_detect(Var2,str_c(groups, collapse = "|"))) %>% 
  dplyr::select(Category=Var2, N=Freq) %>% 
  droplevels() %>% 
  separate(col = N,into = c("Group","N"), sep = ":") %>% 
  dplyr::mutate(N=as.numeric(N)) %>% 
  dplyr::filter(N > 0)
```



```{r}
create_dt(samples_per_category_df)
```

### Number of sequences per sample

```{r}

taxon_table <- feature_table

# Create a phyloseq object
ps <- phyloseq(otu_table(taxon_table, taxa_are_rows = TRUE),
               sample_data(metadata))

sequences_per_sample <- vector2dataframe(named_vector = colSums(otu_table(ps)),
                 column_name = "Sequences_per_sample") %>%
  rownames_to_column("Sample") 


sequences_per_sample %>%
  create_dt()
```

Alpha Plots {data-orientation=rows}
=====================================

Row
-------------------------------------



```{r include=FALSE}
microbe_rarefaction_rds <- glue::glue("{plot_dir}/microbe_rarefaction.rds") 
if(file.exists(microbe_rarefaction_rds)){
  microbe_rarefaction <- readRDS(microbe_rarefaction_rds)
}else{
  # Rarefy the samples to even depth and plot the rarefaction curves
  microbe_rarefaction <- rarefaction(ps= ps, depth = 43847, with_replace = TRUE,
                                     col=custom_palette)
  saveRDS(object = microbe_rarefaction, file = microbe_rarefaction_rds)
}
samples <- sample_data(microbe_rarefaction$ps.rarefied) %>% rownames()
names(samples) <- samples

rarefaction_table <- map2_dfr(.x = seq_along(microbe_rarefaction$rare_curve),
                              .y = samples,
                              .f = function(index=.x, sample=.y){
                                get_rarefaction_df(sample_index =index, 
                                                   sample = sample,
                                                   rare_curve_list=microbe_rarefaction$rare_curve)})

# Add columns for groups to be colored by 
rarefaction_table <- rarefaction_table %>% left_join(metadata, by=join_by(sample == Sample_name))
  # bind_cols(map_dfc(.x = groups,
  #                   .f = map_sample_to_category,
  #                   sample_vector=rarefaction_table$sample,
  #                   metadata=metadata)) %>% droplevels

rarefaction_table_plot <- ggplot(data = rarefaction_table,
                                 mapping = aes(x=sequences, y=observed_species,
                                               color=sample)) + 
  geom_point() + 
  geom_line() + 
  labs(x="Number of sequences", y="Observed OTUs") + 
  publication_format + 
  scale_color_manual(values = custom_palette) + 
  theme(legend.position = "none")

group_plots <- map(groups, .f = function(group){
  
  
  summary_table <- desc_statby(rarefaction_table %>% 
                                 filter(!!sym(group) != "NONE"), 
                               measure.var = "observed_species",
                               grps = c(group,"sequences"))
  summary_table <- pavian::zero_if_na(summary_table)
  
  if(!is.null(annotation_colors[[group]])){
    
    colours <- annotation_colors[[group]]
    
  }else{
    
    colours <- custom_palette
    
  }
  
  ggplot(data = summary_table, mapping = aes_string(x="sequences", y="mean",
                                                    color=group)) + 
    geom_point() + 
    geom_line() + 
    geom_errorbar(aes(ymin=mean-se, ymax=mean+se), width=0.2) +
    labs(x="Number of sequences", y="Observed OTUs") +
    publication_format + 
    scale_color_manual(values = colours)
  
})

group_plots[['samples']] <- rarefaction_table_plot 
```

### Group rarefaction Curves

```{r fig.align='center', fig.height=6, fig.width=6}
# p <- wrap_plots(group_plots, ncol = 2)

ggplotly(group_plots$Treatment)
```

### Samples rarefaction Curves

```{r}
ggplotly(rarefaction_table_plot)
```


Row
-------------------------------------



```{r}

alpha <- map(.x = groups, .f = function(group=.x){
  
  # Estimate alpha diversity
  alpha_diversity(ps.rarefied = microbe_rarefaction$ps.rarefied, 
                  metadata = subset(x = metadata, select = group), 
                  group = group, metrics=c("Observed", "Shannon", "InvSimpson"))
}
)

group <- groups[1]

```

```{r}

# Any of the groups Diversity table can be used since the samples are the same but just grouped differently
diversity.df <- alpha[[group]]$diversity.df %>% rownames_to_column(var = "Samples") %>% 
  dplyr::select(-starts_with("ACE")) %>% 
  dplyr::select(-ends_with("ACE")) %>% 
  dplyr::select(-!!sym(group)) %>% # Drop the group column that will be added from the metadata
  mutate_if(.predicate = is.numeric,.funs = round, digits=2) %>% 
  as.data.frame(check.names=FALSE)
rownames(diversity.df) <- diversity.df[,"Samples"]
diversity.df <- diversity.df[,-(which(colnames(diversity.df) == "Samples"))]


diversity.df <- cbind(metadata[rownames(diversity.df),],diversity.df) 

diversity.df <- diversity.df %>% 
  dplyr::select(-Sample.ID, -se.chao1)  %>% 
  dplyr::select(Sample_name,everything())

# Exclude the control samples from further analysis
no_control_diversity <- diversity.df %>% 
  dplyr::select(!!groups,
                Observed:InvSimpson) 
#%>%dplyr::slice(1:58)

diversity_long.df  <- pivot_longer(data =no_control_diversity, 
                                   cols = Observed:InvSimpson,
                                   names_to = "measures", values_to = 'value')

diversity_metrics <- unique(diversity_long.df$measures)
names(diversity_metrics) <- diversity_metrics 

```

```{r}

terms_formula <- str_c(main_groups, collapse = "*") #main_groups
diversity_anova_results <- map(.x = diversity_metrics, 
                               .f = function(metric=.x, terms=terms_formula, 
                                             response="value"){
  
  factorial_anova(formula = reformulate(termlabels = terms,
                                        response = 'value'),
                  Data = subset(diversity_long.df, measures == metric) ,
                  y = "value", is_one_predictor = TRUE, letters = TRUE)
})
```


```{r}
alpha$Treatment$plot$data$Treatment <- factor(x = alpha$Treatment$plot$data$Treatment,
                                              levels = Treatment_order)

gg <- ggplot(data = alpha$Treatment$plot$data,
             aes(x=Treatment, y=value)) + 
  geom_point() +
  geom_boxplot() + 
  facet_wrap(~variable, scales = "free_y") + 
  labs(x= '', y="Alpha Diversity Measure") +
  publication_format + theme(axis.text.x= element_text(angle = 0))
```


```{r}
alpha$Treatment$plot$data$Treatment  <- factor(x = alpha$Treatment$plot$data$Treatment,
                                               levels = Treatment_order)

df <- alpha$Treatment$plot$data

metrics <- c("Observed", "Shannon")
names(metrics) <- metrics
color_factor <-"Treatment" 
y <- "value"
x <- "Treatment"
#facet_by <- "Soil"
yLAB <- "Alpha Diversity Measure"
color_legend_title <- "Treatment"
term <- "Treatment"

gg <- map(.x = metrics, .f = function(metric=.x){ 
  
  group_box_plot(df =  df %>% filter(variable == metric),
                 x = x, y = y, groups = groups, 
                 palette = annotation_colors$Treatment, 
                 x_lab = "", y_lab = metric,
                 box_color = color_factor,
                 letters_vec = diversity_anova_results[[metric]]$compact_letters[[term]]$Letters)
})
```

### Observed Features

```{r}
ggplotly(gg$Observed)
```

### Shannon

```{r}
ggplotly(gg$Shannon)
```


Alpha Tables
=====================================

Row
-------------------------------------

### Anova Table

```{r}
map_dfr(.x = diversity_metrics,
        .f = function(metric=.x) { 
          diversity_anova_results[[metric]]$anova_test %>%
            dplyr::mutate(Metric=metric) %>%
            dplyr::select(Metric, everything()) }) %>%
  dplyr::mutate(across(where(is.numeric), round, digits=3)) %>%
  create_dt()
```


### Tukey HSD

```{r}
term_names <- names(diversity_anova_results$Observed$compact_letters)
names(term_names) <- term_names

map_dfr(.x = term_names, function(term_name=.x){

  map_dfc(.x = diversity_metrics, .f = function(metric=.x){

    vector2dataframe(named_vector = diversity_anova_results[[metric]]$compact_letters[[term_name]]$Letters,
                     column_name = metric)
  }) %>% rownames_to_column("Group") %>% mutate(Term=!!term_name) %>% dplyr::select(Term,everything())

}) %>% create_dt()

```


Row
-------------------------------------

### Treatment comparison

```{r}
#  Find where there are significant differences between treatments
p_adj <- "none"
group <- "Treatment"
y <- "value"

# Kruskal wallis and Annova test with assumptions testing
map_dfr(.x = diversity_metrics, .f =  function(metric=.x){
  
  anova_analysis(x=group, y = y,
                 Data = subset(diversity_long.df, measures == metric) %>%
                   as.data.frame()) %>% 
    dplyr::mutate(Metric=metric) %>% 
    dplyr::select(Metric, everything()) %>% 
    dplyr::mutate(across(where(is.numeric), round, digits=3))
  
}) %>% create_dt()

```



```{r}

# necessary for estimating fold change
mean_diversity_measure <- summarise_groups(diversity_long.df, 
                                           diversity_metrics,"measures", 
                                           group, "value")
# Get the names of treatments with less then two replicates
# Pairwise comparison between levels
comp <- map_dfr(.x = diversity_metrics, .f = function(metric=.x){
  
  data2use <-  subset(diversity_long.df, measures ==  metric) %>% 
    as.data.frame(check.names=FALSE)
  pairwise_test(Data=data2use, group=group, y_col="value",
                metric=metric, method="Dunn")
  
}
)

# Compact letters display
comp_letters <- map_dfc(.x = diversity_metrics, function(metric=.x){
  
  sub_comp <- comp %>% filter(Metric == metric)
  p_values <- sub_comp$p
  names(p_values) <- paste(sub_comp$group1,sub_comp$group2, sep = "-")
  vector2dataframe(multcompView::multcompLetters(p_values)$Letters, metric)
}) %>% rownames_to_column(group)

group1 <- "Treatment"


mean_se_df <- diversity.df %>% 
  dplyr::select(group1, Observed:InvSimpson) %>%
  dplyr::group_by(!!sym(group1)) %>% 
  dplyr::summarise(N=n(),across(Observed:InvSimpson,list(mean = mean, SE = SE), .names = "{.col}_{.fn}")) %>% 
  dplyr::mutate(across(where(is.numeric), round, digits=2)) %>% 
  as.data.frame()

# Add row names to represent the treatments so they can be tracked
# Here i get the first character of ech group so it agrees with the treatment names used


ROW_NAMES <- mean_se_df[,group1]
rownames(mean_se_df) <- ROW_NAMES
#new_names <- toTitleCase(c(group1,group2,group3) %>%  str_replace_all("_|\\.", " "))

diversity_table <- map_dfc(.x = diversity_metrics, .f = function(metric=.x){
  
  meatric_mean_se <- mean_se_df %>% 
    ungroup() %>% 
    dplyr::select(grep(pattern = metric, x = colnames(.))) %>% 
    as.data.frame()
  
  rownames(meatric_mean_se) <- ROW_NAMES
  
  rownames(comp_letters) <- comp_letters[,group]
  
  combine_mean_se_and_letters(meatric_mean_se, 
                              comp_letters[,metric, drop=FALSE],
                              colname = metric)
  
}) %>% as.data.frame(checknames=FALSE)

rownames(diversity_table) <- comp_letters[,group]

N <- 1+length(groups)
diversity_table <- mean_se_df[rownames(diversity_table),1:N] %>%
  bind_cols(diversity_table)


```

### Treatment table

```{r}
diversity_table %>% 
  create_dt()
```


Beta diversity {data-orientation=rows}
=====================================

Row
-------------------------------------

```{r}
use_qiime <- FALSE

if(use_qiime){
bray_res <- read_qza(bray_file)$data
}else{
bray_res <- create_distance_object(microbe_rarefaction$ps.rarefied) 
}
p <- make_pcoa(bray_res, metadata, color_factor, annotation_colors)
```

### 2D PCOA

```{r}
ggplotly(p)
```

### 3D PCOA

```{r}
plot_3D_pcoa(distance_res = bray_res, metadata = metadata,
             axis1 = "PC1", axis2 = "PC2", axis3 = "PC3",
             color_factor=color_factor, sample_column='Sample.ID', 
             colors2use=annotation_colors[[color_factor]])

```


Row
-------------------------------------

### PERMANOVA: comparing treatments

```{r}
if(use_qiime){
d <- read_qza(bray_distance_file)$data
}else{
  d <- vegdist(taxon_table %>% t) %>% as.matrix() %>% as.data.frame()
  d <- d[sample_order,] %>% as.matrix()
  
}
map_dfr(.x = c(groups), .f = function(group=.x){
  form <- reformulate(group, response = 'd')
  
  permanova_res <- permanova(mat = feature_table[,sample_order] %>% t,
                             metadata = metadata[sample_order, ], 
                             formula = form, dist_method = "bray")
  permanova_res %>% rownames_to_column(var = "Row") %>% 
    dplyr::mutate(Category=group) %>% 
    dplyr::select(Category, everything()) %>% 
    dplyr::mutate(across(where(is.numeric), round, digits=3))
  
}) %>% create_dt()
```

### PERMANOVA: pairwise comparisons

```{r}

comp_df_rds <- glue::glue("{plot_dir}/comp_df.rds") 
if(file.exists(comp_df_rds)){
  comp_df <- readRDS(comp_df_rds)
}else{
# it will locat the the distance object from the parent environment
comp_df <- map_dfr(.x = groups, .f = function(group){
  
  pairwise.adonis(x=d,factors=metadata[,group]) %>% as.data.frame() %>% 
    dplyr::mutate(Category=group) %>% 
    dplyr::rename(Comparison=pairs) %>%
    dplyr::select(Category, everything()) %>% 
    dplyr::mutate(across(where(is.numeric), round, digits=3))
})
 saveRDS(object = comp_df, file = comp_df_rds)
}
comp_df %>% create_dt()
```


Abundance Plots {data-orientation=columns}
=====================================

Column {.sidebar data-width=220}
-------------------------------------

```{r sidebar_ui}

selectInput(inputId = "taxon_level",
            label = "Select a taxonomy level", 
            choices = taxon_levels[-1],
            selected = "Phylum",
            multiple = FALSE)


taxon_table <- reactive({
  taxon_tables[[input$taxon_level]][, sample_order]
  })
```


```{r heatmap_parameters}
colours <- colorRampPalette(c('white','red'))(255)
# manually create color range
hmcol <- c("red", "white", "green")
# expand the color range
hmcol <- colorRampPalette(hmcol)(255)

```


Column
-------------------------------------

### Heatmap

```{r}

plotly::renderPlotly({
  

mat2plot <-  pavian::normalize(taxon_table())
#mat2plot <-  pavian::normalize(taxon_tables[[input$taxon_level]][, sample_order])
mat2plot <- remove_rare_features(feature_table = mat2plot[, sample_order],
                                 cut_off_percent = 1/100, by_sample = FALSE)

heatmaply(x = mat2plot,colors = rev(hmcol), scale = "row", 
          labCol=metadata[colnames(mat2plot),group],
          column_text_angle = 45,dendrogram = "row", show_grid=TRUE,
          xaxis_font_size="16px", yaxis_font_size="16px")

})
```


```{r}
# In percentage
thresholds <- c(Phylum=1,Class=1, Order=1, Family=1, Genus=1, Species=1)

# Convert from wide to long format for every treatment group of interest

level_orders <- list(Treatment_order) 

names(level_orders) <- groups

group_rare <- FALSE
remove_unclassified <- FALSE
maximum_number_of_taxa <- 500

group_relAbundace_tbs <- map2(.x = groups, .y = level_orders, 
                              .f = function(group=.x, level_order=.y){
  
  map2(.x = taxon_levels[-1], .y = taxon_tables[-1], 
       .f = function(taxon_level=.x, taxon_table=.y){
    
    
    if(remove_unclassified){
      
      unclassified_rows <- taxon_table %>% rownames() %>%
        grep("class|assigned", x= .)
      taxon_table <- taxon_table[-unclassified_rows,]
      
    }
    
    taxon_table <- as.data.frame(taxon_table %>% t()) 
    taxon_table <- (collapse_samples(taxon_table = taxon_table,
                                     metadata = metadata, group = group, 
                                     convertToRelativeAbundance = TRUE)$taxon_table * 100 )  %>%  
      as.data.frame(check.names=FALSE)
    
    if(ncol(taxon_table) > maximum_number_of_taxa){
      group_rare <- TRUE
    }
    
    if(group_rare){
      taxon_table <- group_low_abund_taxa(taxon_table %>%  
                                            as.data.frame(check.names=FALSE,
                                                          stringAsFactor=FALSE),
                                          threshold =  thresholds[taxon_level])
      group_rare <- FALSE
    }
    
    taxon_table[,group] <- rownames(taxon_table)
    
    
    #change from wide to long format
    taxon_table <- taxon_table %>% 
      gather(key= !!taxon_level, value = "relativeAbundance", -!!group)
    
    taxon_table[,group] <- factor(x = taxon_table[,group], levels = level_order)
    
    return(taxon_table)
    
  })
  
})


```

```{r}
# Make bar plots
y_lab <- "Relative abundance (%)"  
y <- "relativeAbundance"
x <- "Treatment"
names(group) <- group
group_plots <- map2(.x = group, .y = group, 
                    .f = function(group=.x, x_lab=.y){
  # make plot
  map2(.x = group_relAbundace_tbs[[group]], .y = taxon_levels[-1], 
       .f = function(relAbundace_tb=.x, taxon_level=.y){
         
         df  <- relAbundace_tb %>% dplyr::mutate(Treatment=factor(Treatment, levels=Treatment_order))
         
         ggplot(data =  df, mapping = aes_string(x=x, y=y)) +
           geom_col(aes_string(fill = taxon_level)) + 
           publication_format +
           labs(x = x_lab , y = y_lab) + 
           scale_fill_manual(values = custom_palette) +
           labs(x='')
         
       }
  )}
)
```

Column
-------------------------------------

### Bar plot

```{r}
plotly::renderPlotly({
  
taxon_level <- input$taxon_level
ggplotly(group_plots[[group]][[taxon_level]])

})
```


<!-- ANCOM differential abundance -->
<!-- ===================================== -->

<!-- ```{r eval=FALSE} -->
<!-- ggplotly(ancom_results$Phylum$plot) -->
<!-- ``` -->


Summary
===========================================

<!-- Column  -->
<!-- ----------------------------------- -->

<!-- ### Max Failure Month -->

<!-- ```{r eval=FALSE} -->
<!-- valueBox(max(data$fm), -->
<!--          icon = "fa-user" ) -->
<!-- ``` -->

<!-- ### Average Labor cost -->
<!-- ```{r eval=FALSE} -->
<!-- valueBox(round(mean(data$lc), -->
<!--                digits = 2), -->
<!--          icon = "fa-area-chart") -->
<!-- ``` -->

<!-- ### Average Mileage at Failure -->

<!-- ```{r eval=FALSE} -->
<!-- valueBox(round(mean(data$Mileage), digits = 2), -->
<!--          icon = "fa-area-chart") -->
<!-- ``` -->

Column
-----------------------------------

**Highlights**

* Treatment did not have a significant effect on microbiome community richness and diversity 

* Treatment had a significant effect on the microbiome community structure with major differences usually between the control (CS) and other treatments 

* All treatments were dominated by members of the phylum *Acidobacteria*, *Chloroflexi*, *Plantomycetes* and *Verrucomicrobia*

* *Crenarcheota* was the most dominant Archaea phylum regardless of treatment.



About Report
========================================

**Created by: Olabiyi Obayomi for Jeff Brady's Lab.**

**Institution: Texas A&M AgriLife Research, Texas A&M University.**

This report was generated on `r format(Sys.Date(), format = "%B %d, %Y")`.

